src/Modules/OrchardCore.Commerce/Services/PriceService.cs

```diff
@@ -25,17 +25,41 @@ public class PriceService : IPriceService
 
     public async Task<IList<ShoppingCartItem>> AddPricesAsync(IList<ShoppingCartItem> items)
     {
-        var providers = await _priceProviders
-            .OrderBy(provider => provider.Order)
-            .WhereAsync(provider => provider.IsApplicableAsync(items));
+        var priceProviders = _priceProviders.OrderBy(provider => provider.Order).ToList();
 
-        foreach (var priceProvider in providers)
+        // First let's check for providers that are applicable for all items in the list.
+        var fullyApplicableProviders = await priceProviders.WhereAsync(provider => provider.IsApplicableAsync(items));
+        foreach (var provider in fullyApplicableProviders) items = await provider.UpdateAsync(items);
+
+        // If all applicable providers were used, then there is nothing left to do.
+        var remainingProviders = priceProviders.WhereNot(fullyApplicableProviders.Contains).ToList();
```

Comment:
Hmm can't you use `Except` here? https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.except?view=net-7.0

`priceProviders.Except(fullyApplicableProviders).ToList()`

Reply:
wow, I've never heard about this. Very cool.